core/dbt/adapters/base/impl.py

```diff
@@ -94,6 +91,53 @@ def _utc(dt, source, field_name):
         return dt.replace(tzinfo=pytz.UTC)
 
 
+class SchemaSearchMap(dict):
+    """A utility class to keep track of what information_schema tables to
+    search for what schemas
+    """
+    def add(self, relation):
+        key = relation.information_schema_only()
+        if key not in self:
+            self[key] = set()
+        self[key].add(relation.schema.lower())
+
+    def search(self):
+        for information_schema_name, schemas in self.items():
+            for schema in schemas:
+                yield information_schema_name, schema
+
+    def schemas_searched(self):
+        result = set()
+        for information_schema_name, schemas in self.items():
+            result.update(
+                (information_schema_name.database, schema)
+                for schema in schemas
+            )
+        return result
+
+    def flatten(self):
+        new = self.__class__()
+
+        database = None
+        # iterate once to look for a database name
+        seen = {r.database.lower() for r in self if r.database}
+        if len(seen) > 1:
+            dbt.exceptions.raise_compiler_error(
+                'flatten() requires <=1 database (got {})'.format(seen)
```

Comment:
It looks like this is only used in the pg/redshift version of `_get_cache_schemas`. Can we catch this error and re-raise something pg/redshift-specific? As it stands, I think users would have a hard time knowing what this error means.  Alternatively, we can just call `search()` from the pg's `_get_cache_schemas` and validate for dupes there.

I'd like to show an error message like:
```
Cross-db references not allowed in {adapter} ({db1} vs {db2})
```

I don't feel strongly about the exact wording or location of the exception, but would prefer that it says something like "your database won't let you do this".

The other option is to just call `search()` in pg/redshift and let some downstream SQL handle this. Either something we write (like our `get_catalog()` pg implementation which does not allow multiple dbs), or the database will reject the query with a more contextful error message. Lmk what you think

Reply:
I think my latest commit resolves this, it does something like: `Cross-db references not allowed in {adapter}: Got {seen}`. Does that seem ok?

Reply:
perfect 👍